### PR TITLE
Update xamarin.android and xamarin.ios bindings for handled error attachment API

### PR DESCRIPTION
**Pull request by dhei**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [ ] Are tests passing locally?
* [x] Are the files formatted correctly?
* [ ] Did you add unit tests if this modifies the Windows code?
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Update xamarin.android and xamarin.ios bindings for handled error attachment API

## Related PRs or issues

[AB#71071](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/71071)

#1170 depends on this PR to pass Xamarin CI.

### CODE REVIEW

File: SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Android/Crashes.cs [hunk 1]
```diff
@@ -86,9 +86,9 @@ static Task<bool> PlatformHasReceivedMemoryWarningInLastSessionAsync()
             return Task.Run(() => (bool)future.Get());
         }
 
-        static void PlatformTrackError(Exception exception, IDictionary<string, string> properties)
+        static void PlatformTrackError(Exception exception, IDictionary<string, string> properties, params ErrorAttachmentLog[] attachments)
```
Comment (guperrot) [inline SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Android/Crashes.cs]:
```suggestion
        static void PlatformTrackError(Exception exception, IDictionary<string, string> properties, ErrorAttachmentLog[] attachments)
```

The optional parameters needs to be only in the public layer.

File: SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Android/Crashes.cs [hunk 2]
```diff
@@ -86,9 +86,9 @@ static Task<bool> PlatformHasReceivedMemoryWarningInLastSessionAsync()
...
         {
-            WrapperSdkExceptionManager.TrackException(GenerateModelException(exception, false), properties);
+            WrapperSdkExceptionManager.TrackException(GenerateModelException(exception, false), properties, new ArrayList(attachments));
```
Comment (guperrot) [inline SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Android/Crashes.cs]: This is not null safe

File: SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.iOS/Crashes.cs [hunk 3]
```diff
@@ -73,14 +73,24 @@ static void PlatformNotifyUserConfirmation(UserConfirmation confirmation)
             MSCrashes.NotifyWithUserConfirmation(iosUserConfirmation);
         }
 
-        static void PlatformTrackError(Exception exception, IDictionary<string, string> properties)
+        static void PlatformTrackError(Exception exception, IDictionary<string, string> properties, params ErrorAttachmentLog[] attachments)
```
Comment (guperrot) [inline SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.iOS/Crashes.cs]:
```suggestion
        static void PlatformTrackError(Exception exception, IDictionary<string, string> properties, ErrorAttachmentLog[] attachments)
```

File: SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.iOS/Crashes.cs [hunk 4]
```diff
@@ -73,14 +73,24 @@ static void PlatformNotifyUserConfirmation(UserConfirmation confirmation)
...
         {
-            if (properties != null)
+            var attachmentArray = new NSMutableArray();
```
Comment (guperrot) [inline SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.iOS/Crashes.cs]: Don't allocate a new array if null.

File: SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.iOS/Crashes.cs [hunk 5]
```diff
@@ -73,14 +73,24 @@ static void PlatformNotifyUserConfirmation(UserConfirmation confirmation)
...
+                        attachmentArray.Add(attachment.internalAttachment);
+                    }
+                    else
+                    {
+                        AppCenterLog.Warn(LogTag, "Skipping null ErrorAttachmentLog in Crashes.TrackError.");
+                    }
+                }
             }
-            MSCrashes.TrackModelException(GenerateiOSException(exception, false));
+            MSCrashes.TrackModelException(GenerateiOSException(exception, false), StringDictToNSDict(properties), attachmentArray);
```
Comment (guperrot) [inline SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.iOS/Crashes.cs]: Is StringDictToNSDict handling null already?

File: SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.iOS/Crashes.cs [hunk 6]
```diff
@@ -73,14 +73,26 @@ static void PlatformNotifyUserConfirmation(UserConfirmation confirmation)
...
+                attachmentArray = new NSMutableArray();
+                foreach (var attachment in attachments)
+                {
+                    if (attachment.internalAttachment != null)
+                    {
+                        attachmentArray.Add(attachment.internalAttachment);
+                    }
+                    else
+                    {
+                        AppCenterLog.Warn(LogTag, "Skipping null ErrorAttachmentLog in Crashes.TrackError.");
```
Comment (guperrot) [inline SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.iOS/Crashes.cs]: Xamarin.Android is not logging it, might want to align.

File: SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Android/Crashes.cs [hunk 7]
```diff
@@ -86,9 +86,26 @@ static Task<bool> PlatformHasReceivedMemoryWarningInLastSessionAsync()
...
-            WrapperSdkExceptionManager.TrackException(GenerateModelException(exception, false), properties);
+            ArrayList attachmentArray = null;
+            if (attachments != null)
+            {
+                attachmentArray = new ArrayList();
+                foreach (var attachment in attachments)
+                {
+                    if (attachment?.internalAttachment != null)
+                    {
+
```
Comment (dhei) [inline SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Android/Crashes.cs]:
```suggestion
```

File: Apps/Contoso.Forms.Demo/Contoso.Forms.Demo/ModulePages/CrashesContentPage.xaml.cs [hunk 8]
```diff
@@ -242,9 +242,7 @@ private void TrackException(Exception e)
             }
             Properties.Clear();
             RefreshPropCount();
-            // TODO: uncomment this when API will be added.
-            Crashes.TrackError(e, properties /*, App.GetErrorAttachments().ToArray() */);
-
+            Crashes.TrackError(e, properties, App.GetErrorAttachments().ToArray();
```
Comment (guperrot) [inline Apps/Contoso.Forms.Demo/Contoso.Forms.Demo/ModulePages/CrashesContentPage.xaml.cs]: How was this tested, it would not build without the last `)`.

Comment (michaelsmitha) [reply Apps/Contoso.Forms.Demo/Contoso.Forms.Demo/ModulePages/CrashesContentPage.xaml.cs]: I think he made this change before heading to lunch with the plan of testing after lunch and missed it. We've done some amount of testing with this correct. He wants to do a little more testing.

**Comment by dhei** [comment]
Tested XF Puppet app on iOS and Android with the following scenarios:
- handled error with no property or attachments
- handled error with 1 property
- handled error with 1 property, 1 text attachment
- handled error with 1 property, 1 binary attachment
- handled error with 1 text attachment, 1 binary attachment
- handled error with 1 property, 1 text attachment, 1 binary attachment

File: SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Android/Crashes.cs [hunk 9]
```diff
@@ -86,9 +86,25 @@ static Task<bool> PlatformHasReceivedMemoryWarningInLastSessionAsync()
...
+        static void PlatformTrackError(Exception exception, IDictionary<string, string> properties, ErrorAttachmentLog[] attachments)
         {
-            WrapperSdkExceptionManager.TrackException(GenerateModelException(exception, false), properties);
+            ArrayList attachmentArray = null;
+            if (attachments != null)
+            {
+                attachmentArray = new ArrayList();
+                foreach (var attachment in attachments)
+                {
+                    if (attachment?.internalAttachment != null)
```
Comment (jaelim-ms) [inline SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.Android/Crashes.cs]: Can you invert the condition?

File: SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.iOS/Crashes.cs [hunk 10]
```diff
@@ -73,14 +73,22 @@ static void PlatformNotifyUserConfirmation(UserConfirmation confirmation)
...
+        static void PlatformTrackError(Exception exception, IDictionary<string, string> properties, ErrorAttachmentLog[] attachments)
         {
-            if (properties != null)
+            NSDictionary propertyDictionary = properties != null ? StringDictToNSDict(properties) : new NSDictionary();
+            NSMutableArray attachmentArray = new NSMutableArray();
+            foreach (var attachment in attachments)
             {
-                MSCrashes.TrackModelException(GenerateiOSException(exception, false), StringDictToNSDict(properties));
-                return;
+                if (attachment?.internalAttachment != null)
```
Comment (jaelim-ms) [inline SDK/AppCenterCrashes/Microsoft.AppCenter.Crashes.iOS/Crashes.cs]: Invert the condition.

**Pull request merged**